app/controllers/posts_controller.rb

```diff
@@ -1,19 +1,64 @@
 class PostsController < ApplicationController
 
+  before_action :get_posts, only: [:index, :show]
+  before_action :fetch_post, only: [:show, :edit, :update, :destroy]
+  before_action :get_ad
```

Comment:
You have this in both controllers. They both descend from ApplicationController. Why not have the `get_ad` method and the before_action there?